Use `likes` and `shares` totalItems on status creations and updates

sneakers-the-rat opened the pull request:
"the likes and the boosts are wrong" is a super common complaint. 

Reading previous discussion, it seems like choosing to not enumerate all favourites and boosts was an active decision, and that seems fine. 

I don't see a similar reason why not to use the like/share counts when they are provided by the server tho.

When I fetch a status like this: 

```shell
curl -H 'Accept: application/ld+json' https://neuromatch.social/@jonny/113347217701900056
```

I get a JSON-LD object like this:

```json
{
    "id": "https://neuromatch.social/users/jonny/statuses/113347217701900056",
    "type": "Note",
    "...": "...",
    "likes":
    {
        "id": "https://neuromatch.social/users/jonny/statuses/113347217701900056/likes",
        "type": "Collection",
        "totalItems": 50
    },
    "shares":
    {
        "id": "https://neuromatch.social/users/jonny/statuses/113347217701900056/shares",
        "type": "Collection",
        "totalItems": 39
    }
}
```

Here the originating server is telling me the correct number of boosts and favorites, which should supercede any local count we have. Currently this information is just discarded.

This PR, quite simply, uses that information when it is present.

When combined with https://github.com/mastodon/mastodon/pull/32615 , this means that expanding a status to its detailed view will update the boost and favorite count for a status and all its replies. 

There is some possibility for mismatches, and these counts will also get overridden if someone runs the `tootctl` command to update counts, but in either case I think it's still desirable to use "ground truth" information when we get it.

I think the abuse potential here is super low. I can't think of an abuse scenario when knowing an accurate boosts/faves count (which could be gotten by any logged out user by visiting the post's page) would cause harm - recall that fetch all replies would be the main trigger for this for already-existing statuses, and it doesn't fetch any followers-only or dm statuses.

If this looks acceptable, i will update https://github.com/mastodon/mastodon/pull/32615 to also fetch the root post in order to update its stats

sneakers-the-rat committed:
use activitypub counts when updating or creating a status

sneakers-the-rat committed:
use status parser instead

sneakers-the-rat committed:
tests

Gargron commented:
Trending posts are calculated based on these numbers, so the abuse potential is allowing someone to get on top of the explore page with fake numbers. I’d suggest created untrusted counter columns to store this information and then only use them for display purposes.

sneakers-the-rat committed:
add untrusted counters to status stats

sneakers-the-rat committed:
oop we actually want untrusted favourites

sneakers-the-rat committed:
use untrusted values in serializer but not in trends calculations, add tests

sneakers-the-rat committed:
increment and decrement in between getting true values

sneakers-the-rat commented:
Alright, updated to use `untrusted_reblogs` and `untrusted_favourites` columns instead of the existing columns.

- Separate columns
- That have their values assigned/overridden by ActivityPub create/update events
- And are incremented/decremented along with reblogs/favourites in between those events

Added tests:
- `Rest::StatusSerializer` - use the untrusted values when they are present, otherwise fallback to normal values
- `Status` - Increment/decrement on reblog/favourite
- `Trends` - untrusted values are ignored when calculating trends

Some notes - so i don't repeat myself, imagine i said "let me know if i should change this" to all of them:
- There weren't existing tests that I could find for the `StatusSerializer`, and I kept getting an error from `current_user` being undefined. I search around in the tests and online to see how that helper works and couldn't find an answer, so I added a check for `current_user` being defined, which seems like it preserves the existing semantics of "is there a current user." 
- I made these columns nullable to indicate absence of information rather than 0, and made them always `nil` for local statuses. Wasn't sure how/whether it would be necessary to drop assigned values for local statuses, that would mostly be for db size, but probably negligible.
- I used `integer` but noticed that the other columns use `bigint` - I looked up the limits, and it seemed like 4 billion was big enough?
- I added a clamp to the values to keep them in the range of 0 and 1 billion (may we be so bold to dream of this ever happening on the fedi), i didn't bother to pull that value out to a classvar & figured that would avoid needing to make accessor methods that return a positive int, but i also can follow that pattern if that's better.

Gargron commented on app/models/status.rb:
```diff
@@ -303,12 +303,26 @@ def favourites_count
+
   def increment_count!(key)
     update_status_stat!(key => public_send(key) + 1)
+    increment_count!(:untrusted_favourites_count) if key == :favourites_count && !untrusted_favourites_count.nil?
```
This will issue up to 3 different `UPDATE`s from a single method call, I'm also uncomfortable with the recursivity here.

sneakers-the-rat replied:
gotcha. wrote it this way just to try and minimize changes to the method, but i can rewrite it to batch

Gargron replied:
At the very least this could be:

```ruby
def increment_count!(key)
  if key == :favourites_count && !untrusted_favourites_count.nil?
    update_status_stat!(favourites_count: favourites_count + 1, untrusted_favourites_count: untrusted_favourites_count + 1)
  elsif key == :reblogs_count && !untrusted_reblogs_count.nil?
    # Equivalent...
  else
    update_status_stat!(key => public_send(key) + 1)
  end
end
```

sneakers-the-rat replied:
ah i just did this here https://github.com/mastodon/mastodon/pull/32620/commits/29729167b77d728682ada4c79e18a68ef59ec768

but i like that better, one sec

sneakers-the-rat replied:
that's here - https://github.com/mastodon/mastodon/pull/32620/commits/fcd970170e42cceccf5d9c69471af5cceb0a1d5c

doesn't feel very DRY but i think moving those methods to the counter model/making a counter concern would be the next step if we expected linked values like this to happen again

sneakers-the-rat commented on app/serializers/rest/status_serializer.rb:
```diff
@@ -45,7 +45,7 @@ def in_reply_to_account_id
 
   def current_user?
-    !current_user.nil?
+    defined?(current_user) && !current_user.nil?
```
great! i will look around for how to do that, just didn't know what to search for since the link is sufficiently abstract to me (not used to ruby)

sneakers-the-rat replied:
done in https://github.com/mastodon/mastodon/pull/32620/commits/22ed008e4c8f9b92bbe550a80f671a88dc067b95

modified the way options were passed to serializers in helpers to avoid a bunch of special case kwargs, should be equivalent by default

sneakers-the-rat committed:
extract max untrusted cont to a class const

sneakers-the-rat committed:
use bigint rather than integer, raise cap to 100m

sneakers-the-rat committed:
use scope and scope_name in serializer spec

sneakers-the-rat committed:
update counts with one update call rather than 3

sneakers-the-rat committed:
clearer if/elsif increment/decrement methods

Gargron commented on app/models/status.rb:
```diff
@@ -393,6 +415,9 @@ def unlink_from_conversations!
     return if marked_for_destruction? || destroyed?
 
+    attrs[:untrusted_favourites_count] = attrs[:favourites_count] if attrs.key?(:favourites_count) && !untrusted_favourites_count.nil?
+    attrs[:untrusted_reblogs_count] = value if key == :favourites_count && !untrusted_favourites_count.nil?
```
I think `value` is a typo here.

sneakers-the-rat replied:
oops leftovers from a prior attempt, reverting

sneakers-the-rat committed:
revert update_status_stat leftovers

trwnh commented:
the use of untrusted numbers even for display makes me uneasy and i think there should be an option or at least some way to show only trusted numbers. at the very least, any time an untrusted number is shown, it should be made clear that the number is untrusted, either by adding a flag/asterisk somewhere saying "this number is untrusted", or by also including the trusted number alongside it like "2134 likes (12 verified)"

related #11339 and https://github.com/mastodon/mastodon/issues/11339#issuecomment-1465098738

sneakers-the-rat commented:
What is the risk scenario here? Are we worried about people spoofing high numbers or low numbers? 

The thing this is trying to address is that there is a persistent annoyance at a number of related things about the fediverse, and mastodon in particular, that it's confusing that what they see is different from everyone else. counts on likes and boost is one of the smaller ones, but it comes up often.

from what i read of #11339 and understand of AP generally, getting an accurate global count of interactions is not really possible at the moment, so using the numbers that the origin servers tell us about is the next best thing.

I do expect some instances to just put in fake numbers here because it would be funny, but it would also sort of screw up paging for them since the count given on the `OrderedCollection` would be super wrong so it's at their peril. 

The UX here should be unchanged - if you click on the counts, you still see the same number of favorites/boosts and only those your instance knows about. 

we could do something like this:
<img width="697" alt="Screenshot 2024-10-28 at 10 08 36 PM" src="https://github.com/user-attachments/assets/ac984f5a-19c3-43bd-9584-19cdd53ef30d">

with an explanatory tooltip when you mouseover it?

I'm fine with adding in-line documentation and all that but i also am not quite sure what the worst case of spoofing interactions is

trwnh commented:
generally it is more socially profitable to spoof higher numbers, without
making it too obvious. so for example, a profile with 30 followers can
claim it has 3000 followers (not an unrealistic number), and you wouldn't
know because the contents of the `followers` collection are not public.
with likes and shares, the same thing can be done for statuses which report
inflated numbers in a mostly undetectable way. for the same reason we
shouldn't use these untrusted numbers to calculate trends, i am arguing
that we should likewise not show them to users without any indication that
the numbers are untrusted. users need to be aware of when they are
potentially being lied to.

you can get a "global count" by asking the origin responsible for hosting
the object what its likes.totalItems or shares.totalItems are, but
authority and trust are not the same thing. it's not just "funny" that we
have to worry about here -- generally, "funny" will try to make itself
obvious because that's part of the joke when you see someone has whatever
the current total population of the earth is as their followers.totalItems,
or someone has 69 likes.totalItems and 420 shares.totalItems on every
single one of their statuses. the real threat here is that someone subtly
inflates numbers, and that we just show this number to users without any
context. it's not clear if the claim is being made by the local instance,
or by the remote instance. users are assumed to trust their local instance,
but we assume that remote instances aren't trusted.

the totalItems count and the paging mechanisms are entirely separate in
most cases. especially here, where the items are private and not pageable
at all. so it doesn't mess up paging to lie, you can just lie with zero
consequences and zero ways to detect the lie. the lie doesn't have to be
malicious, for example you can claim 0 followers 0 likes 0 shares all the
time, out of a desire for privacy perhaps. but the worst case is
disinformation actors subtly manipulating their numbers to make themselves
look more trustworthy, and users taking the numbers at face value because
they were never informed of even the *possibility* that these numbers might
be faked.

UX-wise i don't want to see the untrusted numbers. in most cases i don't
particularly care to see the trusted numbers either unless i ask for them,
but i understand many other people do want to see numbers. i am assuming
right now that both the timeline view and the detailed status view will
show the untrusted number, but clicking through to the likes will show the
people who are verified to have liked it. this is not good enough, because
most people don't click through this far, and even this far will not show
the verified vs unverified distinction. the conscionable thing to do is not
show counts in timelines, or otherwise show something like "⭐ 12k (3)" but
this might be confusing because users won't know what the (3) means. then,
on the detailed status view and on the list of likes, you show something
like "⭐ 11,679 favourites (3 verified)". what would also make sense is an
option which lets users and admins choose one of the following:

- hide counts
- show trusted counts only
- show trusted and untrusted counts

the objection i am raising here is not in fetching or storing the untrusted
numbers -- this is fine, if this was all the pr does. the objection is
about showing these untrusted numbers without clearly disclaiming that they
are untrusted. we have to keep in mind that most users do *not* know the
numbers can be faked, and they do *not* clickthrough on every status in
their timeline. switching to show untrusted numbers is a regression in user
trust.

On Tue, Oct 29, 2024, 00:10 Jonny Saunders ***@***.***> wrote:

> What is the risk scenario here? Are we worried about people spoofing high
> numbers or low numbers?
>
> The thing this is trying to address is that there is a persistent
> annoyance at a number of related things about the fediverse, and mastodon
> in particular, that it's confusing that what they see is different from
> everyone else. counts on likes and boost is one of the smaller ones, but it
> comes up often.
>
> from what i read of #11339
> <https://github.com/mastodon/mastodon/issues/11339> and understand of AP
> generally, getting an accurate global count of interactions is not really
> possible at the moment, so using the numbers that the origin servers tell
> us about is the next best thing.
>
> I do expect some instances to just put in fake numbers here because it
> would be funny, but it would also sort of screw up paging for them since
> the count given on the OrderedCollection would be super wrong so it's at
> their peril.
>
> The UX here should be unchanged - if you click on the counts, you still
> see the same number of favorites/boosts and only those your instance knows
> about.
>
> we could do something like this:
> Screenshot.2024-10-28.at.10.08.36.PM.png (view on web)
> <https://github.com/user-attachments/assets/ac984f5a-19c3-43bd-9584-19cdd53ef30d>
>
> with an explanatory tooltip when you mouseover it?
>
> I'm fine with adding in-line documentation and all that but i also am not
> quite sure what the worst case of spoofing interactions is
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mastodon/mastodon/pull/32620#issuecomment-2443226854>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACQ5OX75D5LRTCP3L2VIG7TZ54KDPAVCNFSM6AAAAABQLVFFH6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDINBTGIZDMOBVGQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

sneakers-the-rat commented:
given that the constraints are "people want to see the numbers" and "it's impossible/impractical to get accurate numbers" I agree there are no good answers here. 

I disagree that it's inherently a regression in user trust, especially if we take the next step (which i think we should) of signaling where the numbers come from as well as the numbers we know for sure. The current numbers are *always already* "lies" especially for smaller instances, it's just the lie goes in the other direction where it seems like there is nobody out there, and one could similarly argue that artificially deflating numbers globally has its own harm story. 

I would argue that the current situation is much less clear re: the origin of the numbers - It's the count of the interactions... of people who someone on my instance follow? I think? in some circumstances? It's not clear whether it's our instance or their instance because there isn't a well defined answer for whose "fault" it is the numbers are wrong. This PR actually clarifies that - one number is the number that the other instance says, the other number is the number we have seen. Maybe 'trusted' and 'untrusted' have too much of a moral valence to them, they could equivalently be called 'attested' and 'observed,' because the local number isn't intrinsically more trustworthy - we know it's always wrong - it just represents the number observed.

It wouldn't be too hard to make a switch if we're going to serialize both in the `RESTSerializer` anyway, that also sounds like a good idea.

trwnh commented:
> Maybe 'trusted' and 'untrusted' have too much of a moral valence to them,
they could equivalently be called 'attested' and 'observed,' because the
local number isn't intrinsically more trustworthy - we know it's always
wrong - it just represents the number observed.

it's less about the number being "wrong" and more about the number
representing the strict subset of interactions that have been verified. in
plain english it would be like saying "the remote claims 12k likes but i'm
only aware of 3". when i say it's a "regression in user trust" i mean that
previously the user could trust that there are *at least* as many
interactions as the number. now they can't.

regarding what to call it, i don't think "trust" is inherently moral, but
if the terminology is a problem then the next best term is "verified" and
"unverified". the reality we are trying to represent here is that there is
no central "truth", there are only "claims", and it matters who is making
the claim.

regarding harm stories, i'm honestly not a fan of emphasizing numbers to
this degree at all, but again i recognize i am in the minority view here. i
do think that the (demonstrable) harms of inflating numbers are much more
significant than the (alleged) harms of showing only a subset. in either
case, more information needs to be presented to the user, because neither
number should be taken at face value without clarifying what that number
represents.

On Tue, Oct 29, 2024, 01:12 Jonny Saunders ***@***.***> wrote:

> given that the constraints are "people want to see the numbers" and "it's
> impossible/impractical to get accurate numbers" I agree there are no good
> answers here.
>
> I disagree that it's inherently a regression in user trust, especially if
> we take the next step (which i think we should) of signaling where the
> numbers come from as well as the numbers we know for sure. The current
> numbers are *always already* "lies" especially for smaller instances,
> it's just the lie goes in the other direction where it seems like there is
> nobody out there, and one could similarly argue that artificially deflating
> numbers globally has its own harm story.
>
> I would argue that the current situation is much less clear re: the origin
> of the numbers - It's the count of the interactions... of people who
> someone on my instance follow? I think? in some circumstances? It's not
> clear whether it's our instance or their instance because there isn't a
> well defined answer for whose "fault" it is the numbers are wrong. This PR
> actually clarifies that - one number is the number that the other instance
> says, the other number is the number we have seen. Maybe 'trusted' and
> 'untrusted' have too much of a moral valence to them, they could
> equivalently be called 'attested' and 'observed,' because the local number
> isn't intrinsically more trustworthy - we know it's always wrong - it just
> represents the number observed.
>
> It wouldn't be too hard to make a switch if we're going to serialize both
> in the RESTSerializer anyway, that also sounds like a good idea.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mastodon/mastodon/pull/32620#issuecomment-2443299898>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACQ5OX7MYGGYLOLO2BVQG4DZ54RNBAVCNFSM6AAAAABQLVFFH6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDINBTGI4TSOBZHA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

sneakers-the-rat commented:
> the user could trust that there are *at least* as many
interactions as the number.

this could also be framed as "the user could never trust that the number they were seeing was accurate" or "the user could be certain that the number they were seeing was inaccurate" and see also how i think this could make things more clear by adding detail in the ux on mouseover.

> more information needs to be presented to the user

yep, no disagreement here.

Gargron commented:
Just to weigh in, the problem already exists for follower/following numbers, so I don't see it as a blocker for showing other federated numbers. I'll run the display options by our designer at the next opportunity.

shleeable commented:
This is one of those things... 99.999% of toots will be "real". 
people can spoof them but if that becomes a problem, it's easy to detect and generally coming from a single bad actor/user.

Not a systemic issue.

trwnh commented:
it's not easy to detect; it's *impossible* to detect. the only way you could possibly suspect anything is if the numbers are disproportionately extreme, which is a threshold that is entirely subjective. is it unreasonable for someone with 3000 followers to get tens of interactions? hundreds? whatever you think is "too many", i could claim a number just slightly under that. what are the chances anyone is going to investigate further? or are they just going to take the numbers at face value and be psychologically influenced by them?

the problem of course gets worse when you realize that i might not even have 3000 followers -- i might be lying about that number as well! so you have no frame of reference on what is "reasonable".

br4yd commented:
If I understand it correctly this has been merged from the technical part but they are not showing up as the federated amount on the user's end, correct?

Nikophos commented:
> If I understand it correctly this has been merged from the technical part but they are not showing up as the federated amount on the user's end, correct?

My reply from @renchap might be worth noting (screenshot below); this PR was merged into the nightly release and won't be included in a published release until v4.4.0.

![MastodonScreenshot_Private](https://github.com/user-attachments/assets/f6585591-4b03-4574-a5d8-041aa9814b14)

---

I think you're correct that only the technical part was implemented. Viewing the post below using my account on mastodon.social (v4.4.0-nightly.2025-03-10),

![image](https://github.com/user-attachments/assets/8b06e1a8-c299-408d-94a3-12d3563509bd)

in comparison to the original mas.to (v4.3.5) page,

![image](https://github.com/user-attachments/assets/60f93220-d888-4657-ad2e-31ff687c2952)

shows that the boosts and favorites appear to incomplete.

---

As a relatively non-technical user, I'd really like to see this change fully implemented. I'm personally in favour of showing both trusted & untrusted numbers like in @sneakers-the-rat's [reply](https://github.com/mastodon/mastodon/pull/32620#issuecomment-2443226854). A potential compromise could be displaying the untrusted number in brackets, instead. That way users could view the untrusted count when they expand the post, and the trusted values would still be the "main" display number (just as they currently are).

On a similar note, I think it would be great if users could choose whether they want the main display number to be the trusted or untrusted count. A toggle in the user Account Settings could work. In that case, I'd agree that "verified" and "unverified" might be more accurate & clear terminology.

br4yd commented:
> I think you're correct that only the technical part was implemented. Viewing the post below using my account on mastodon.social (v4.4.0-nightly.2025-03-10),

Interestingly I am on glitch-soc v4.4.0-alpha.4+glitch which then should have this already implemented, but can't see federated likes either :thinking: or it's not merged into glitch-soc itself yet even though it's basically using v44.4.0-alpha from upstream Mastodon?

sneakers-the-rat commented:
Hmm, you're right. just tried this locally and the `untrusted_` counts are null for me, even though the appropriate fields are present in the source JSON-LD (and when i initially wrote this i confirmed this worked and wrote tests that i thought should have protected it). will investigate
